cpp/ql/lib/semmle/code/cpp/dataflow/ExternalFlow.qll

```diff
@@ -0,0 +1,540 @@
+    i = 2 * n and result = p
+    or
+    i = 2 * n - 1 and result = "," and n != 0
+  )
+  or
+  i = 2 * c.getNumberOfParameters() and result = ")"
+}
+
+/**
+ * Gets a parenthesized string containing all parameter types of this callable, separated by a comma.
+ *
+ * Returns the empty string if the callable has no parameters.
+ * Parameter types are represented by their type erasure.
+ */
+cached
+string paramsString(Function c) { result = concat(int i | | paramsStringPart(c, i) order by i) }
+
+bindingset[func]
+predicate matchesSignature(Function func, string signature) {
+  signature = "" or
+  paramsString(func) = signature
+}
+
+/**
+ * Gets the element in module `namespace` that satisfies the following properties:
+ * 1. If the element is a member of a class-like type, then the class-like type has name `type`
+ * 2. If `subtypes = true` and the element is a member of a class-like type, then overrides of the element
+ *    are also returned.
+ * 3. The element has name `name`
+ * 4. If `signature` is non-empty, then the element has a list of parameter types described by `signature`.
+ *
+ * NOTE: `namespace` is currently not used (since we don't properly extract modules yet).
+ */
+pragma[nomagic]
+private Element interpretElement0(
+  string namespace, string type, boolean subtypes, string name, string signature
+) {
+  elementSpec(namespace, type, subtypes, name, signature, _) and
+  namespace = "" and // TODO: Fill out when we properly extract modules.
+  (
+    // Non-member functions
+    exists(Function func |
+      func.getName() = name and
+      type = "" and
+      matchesSignature(func, signature) and
+      subtypes = false and
+      not exists(func.getDeclaringType()) and
+      result = func
+    )
+    or
+    // Member functions
+    exists(Class namedClass, Class classWithMethod, Function method |
+      method.getName() = name and
+      method = classWithMethod.getAMember() and
+      namedClass.getName() = type and
+      matchesSignature(method, signature) and
+      result = method
+    |
+      // member declared in the named type or a subtype of it
+      subtypes = true and
+      classWithMethod = namedClass.getADerivedClass*()
+      or
+      // member declared directly in the named type
+      subtypes = false and
+      classWithMethod = namedClass
+    )
+    or
+    // Member variables
+    signature = "" and
+    exists(Class namedClass, Class classWithMember, MemberVariable member |
+      member.getName() = name and
+      member = classWithMember.getAMember() and
+      namedClass.getName() = type and
+      result = member
+    |
+      // field declared in the named type or a subtype of it (or an extension of any)
+      subtypes = true and
+      classWithMember = namedClass.getADerivedClass*()
+      or
+      // field declared directly in the named type (or an extension of it)
+      subtypes = false and
+      classWithMember = namedClass
```

Comment:
I'm not quite convinced we need these for our own use-cases just yet. Could we delay this part until we have a concrete case where we need this?

Reply:
Supporting derived classes is just what the `subtypes` field does, and it's there, so I feel it should work.  The comments are off though - C++ doesn't have 'extensions'.

Reply:
I mean the whole "member variables" thing. The only query I know that uses a member variable as a source is the `cpp/cleartext-transmission` query and that one has quite a lot of machinery for selecting variables with appropriate names, so I don't think this query will be a candidate for MaD initially.

Thus, I think we should wait with the whole "member variable" case in this predicate.

Reply:
The lack of (or incomplete?) support for variables as sources/sinks is one of the things that frustrated me with the initial Swift MAD implementation, until we fixed it.  I strongly believe leaving this out is just leaving a trap for someone in near future who expects it to 'just work'.

Also this is already tested and, I think, working (see `memberRemoteMadSourceVar` in the test).

Reply:
Alright, that's fair. If you believe this will be useful then let's keep it 👍